### PR TITLE
feat(meta): add Meta Model API provider (OpenAI + Anthropic dual protocol)

### PR DESCRIPTION
**Pull request by tbille**
## Description

Adds a new `meta` provider for [Meta's Model API](https://dev.meta.ai/docs) (`muse-spark-*` models). This API is unusual in that it serves the same models through **both** protocols on the same account/key:

- Chat Completions + Responses via `https://api.meta.ai/v1`, OpenAI-SDK compatible.
- Messages via `https://api.meta.ai`, Anthropic-SDK compatible (native `content` blocks, `thinking`, `tool_use`, `stop_reason`).

`MetaProvider(BaseOpenAIProvider)` inherits Chat Completions and Responses unmodified (no translation needed, they're wire-compatible with the OpenAI SDK), and stands up a second `AsyncAnthropic` client to serve `.messages()`/`.amessages()` as a genuine native pass-through to `/v1/messages`, rather than any-llm's default Messages↔Completions bridge — this preserves `thinking` blocks, native `tool_use` blocks, and `cache_control` that a bridge conversion would otherwise drop. Two auth details worth calling out for review:

- The Anthropic client is constructed with `auth_token=` (not `api_key=`), since Meta expects `Authorization: Bearer` rather than Anthropic's native `x-api-key` header.
- The Anthropic SDK's base URL is derived from the OpenAI one by stripping a trailing `/v1` (`https://api.meta.ai/v1` → `https://api.meta.ai`), since the Anthropic SDK appends `/v1/messages` itself.

`context_management` and `betas` are rejected with `NotImplementedError` on `.messages()`, since Meta's translation layer doesn't document support for Anthropic's context-management/beta primitives.

This lands as **community tier**: added to the `LLMProvider` enum and a `pyproject.toml` extra, but **not** to `VERIFIED_PROVIDERS` or the `tests/conftest.py` model maps, since I don't hold a `MODEL_API_KEY` for CI. Capability flags were set conservatively from the docs alone (no live key to verify against):

- `SUPPORTS_COMPLETION_REASONING=False` — docs state `reasoning_content` on Chat Completions is redacted to empty for external callers.
- `SUPPORTS_EMBEDDING`, `SUPPORTS_BATCH`, `SUPPORTS_IMAGE_GENERATION`, `SUPPORTS_RERANK` all `False` — none of these endpoints are documented.
- `PROMPT_CACHE_KEY_SUPPORT="supported"` — docs confirm `prompt_cache_key` on Chat Completions + Responses (not explicitly confirmed on Messages).

**I could not run this against the live endpoint** (no `MODEL_API_KEY`). Whoever has one, the live-verification snippet from `CONTRIBUTING.md` would be:

```python
from any_llm import AnyLLM
llm = AnyLLM.create("meta", api_key="...")
print(llm.completion(model="muse-spark-1.2", messages=[{"role": "user", "content": "Say hi"}]).choices[0].message.content)
print(llm.messages(model="muse-spark-1.2", max_tokens=64, messages=[{"role": "user", "content": "Say hi"}]))
print([m.id for m in llm.list_models()][:5])
```

That run would also be the moment to correct any of the conservative flags above if the live behavior differs from the docs.

## PR Type

- 🆕 New Feature

## Relevant issues

N/A

## Checklist

- [x] I understand the code I am submitting.
- [x] I have added unit tests that prove my fix/feature works
- [ ] I have run this code locally and verified it fixes the issue. — unit tests pass locally; the live endpoint is not verified (no `MODEL_API_KEY`), see note above.
- [x] New and existing tests pass locally (`uv run pytest tests/unit` — 1889 passed; the only failures are a pre-existing local `otari` SDK version mismatch unrelated to this change, reproduced identically on unmodified `main`)
- [x] Documentation was updated where necessary (`docs/providers.md` is generated from provider metadata; no manual docs needed)
- [x] I have read and followed the [contribution guidelines](https://github.com/mozilla-ai/any-llm/blob/main/CONTRIBUTING.md)
- [x] **AI Usage:**
    - [ ] No AI was used.
    - [x] AI was used for drafting/refactoring.
    - [ ] This is fully AI-generated.

## AI Usage Information

- AI Model used: Claude Sonnet 5
- AI Developer Tool used: Claude Code
- Any other info you'd like to share: Implementation, tests, and this PR description were drafted by Claude Code from the published Meta Model API docs, following the existing `otari`/`anthropic` provider patterns in this repo. Live-endpoint verification is still needed from someone with a `MODEL_API_KEY`.

- [x] I am an AI Agent filling out this form

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for the Meta provider.
  * Supports OpenAI-compatible completions and responses, as well as Anthropic-compatible messaging.
  * Added standard and structured outputs, streaming responses, authentication, and model listing.
  * Included the Meta provider in the complete optional installation.

* **Tests**
  * Added comprehensive coverage for Meta provider configuration, requests, streaming, authentication, and output handling.

### CODE REVIEW

File: src/any_llm/providers/meta/meta.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+
+def _derive_anthropic_base(openai_base: str) -> str:
+    """Derive the Anthropic SDK base URL from the OpenAI SDK one.
+
+    Meta's OpenAI-compatible endpoints live under `.../v1` (the OpenAI SDK appends
+    `/chat/completions`, `/responses`, etc. itself); the Anthropic SDK appends
+    `/v1/messages` itself, so it needs the bare host instead.
+    """
+    return openai_base.removesuffix("/v1")
```
[the remark this answers is not in the document]
Comment (tbille) [reply src/any_llm/providers/meta/meta.py]: Fixed in de41596: normalizes the trailing slash before stripping `/v1` (`openai_base.rstrip("/").removesuffix("/v1")`), and added a test covering a `.../v1/` override.

File: src/any_llm/providers/meta/meta.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            native_kwargs = params.model_dump(
+                exclude_none=True, exclude={"output_format", "stream", "betas", "context_management"}
+            )
+            native_kwargs.update(kwargs)
+            if is_structured_output_type(params.output_format):
+                return await self._anthropic_client.messages.parse(output_format=params.output_format, **native_kwargs)
+            message = await self._anthropic_client.messages.create(
+                output_config=cast("Any", params.output_format), **native_kwargs
+            )
+            return self._convert_native_message_to_response(message)
```
[the remark this answers is not in the document]
Comment (tbille) [reply src/any_llm/providers/meta/meta.py:151]: Fixed in de41596: `_amessages` now raises `ValueError("stream is not supported for output_format")` if both are set, instead of silently dropping `stream`. This mirrors the guard already in the public `amessages()` entry point (any_llm.py), so it's now enforced at both layers. Added a test for the direct-call case.

File: tests/unit/providers/test_meta_provider.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+    types = [event.type for event in collected]
+    assert types == [
+        "message_start",
+        "content_block_start",
+        "content_block_delta",
+        "content_block_stop",
+        "message_delta",
+        "message_stop",
+    ]
```
[the remark this answers is not in the document]
Comment (tbille) [reply tests/unit/providers/test_meta_provider.py]: Fixed in de41596: the test now asserts each collected event is the expected any-llm event class with its key converted fields (message id, content-block type, delta text, stop_reason, usage), not just `event.type`. Also added an unrecognized SDK event (`ping`, which real Anthropic streams do send) to the input list and asserted it's dropped rather than yielded.

File: src/any_llm/providers/meta/meta.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    # reasoning_content is documented as redacted-empty for external callers on Chat
+    # Completions; flip to True once verified against the live endpoint.
+    SUPPORTS_COMPLETION_REASONING = False
+    SUPPORTS_COMPLETION_IMAGE = True
+    SUPPORTS_COMPLETION_PDF = True
+    SUPPORTS_EMBEDDING = False
+    SUPPORTS_MODERATION = False
+    SUPPORTS_LIST_MODELS = True
+    SUPPORTS_BATCH = False
+    SUPPORTS_IMAGE_GENERATION = False
```
[the remark this answers is not in the document]
Comment (tbille) [reply src/any_llm/providers/meta/meta.py:95]: Agreed in principle, and this is already the biggest disclosed gap in the PR description: this is a community-tier provider and I don't hold a `MODEL_API_KEY`, so none of these flags have been exercised against the live endpoint. I've added an explicit code comment above the flag block saying so, so it's visible in the source too, not just the PR description.

I'm intentionally not flipping any of these to `False` pre-emptively though: `SUPPORTS_RESPONSES`/`SUPPORTS_COMPLETION_IMAGE`/`SUPPORTS_COMPLETION_PDF` are all directly and unambiguously stated in Meta's published docs (three endpoints incl. Responses; "Text, image, video, PDF" input modalities), not inferred/guessed — turning them off would just make the provider under-report real capability without adding any actual safety, since the underlying inherited `BaseOpenAIProvider` code paths are identical to every other OpenAI-compatible provider in this repo. `SUPPORTS_COMPLETION_REASONING` is the one flag I did set conservatively (`False`) specifically because the docs describe a caveat (redacted-empty `reasoning_content` for external callers) that would make the flag misleading if left on.

Whoever has a key: the live-verification snippet is in the PR description (mirrors CONTRIBUTING.md's format) — running it is the real fix here, not a code change I can make blind.

File: tests/unit/providers/test_meta_provider.py [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+        messages=[{"role": "user", "content": "Hello"}],
+        max_tokens=64,
+        output_format=output_config,
+    )
+    result = await provider._amessages(params)
+
+    assert isinstance(result, MessageResponse)
+    call_kwargs = provider._anthropic_client.messages.create.call_args.kwargs
+    assert call_kwargs["output_config"] == output_config
+
```
[the remark this answers is not in the document]
Comment (tbille) [reply tests/unit/providers/test_meta_provider.py:373]: Declining this one: `MetaProvider` doesn't override `_aresponses` at all — Responses is 100% inherited, unmodified `BaseOpenAIProvider._aresponses`, including its `parse_responses_output`/dataclass-dict path and the Pydantic `responses.parse()` path referenced in the coding guideline. Both are already exercised by the shared base-provider test suite (e.g. `tests/unit/providers/test_openai_base_provider.py`) against that same shared code. Adding a Meta-specific copy of those tests would just re-run the same assertions against a different subclass instance with zero new branches to cover — no new code in this PR touches Responses structured output.

Happy to add it if a maintainer would rather have the redundant per-provider coverage anyway, but I don't think it earns its keep here.

File: src/any_llm/providers/meta/meta.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+        if params.output_format is not None:
+            if params.stream:
+                # Belt-and-suspenders: the public `amessages()` entry point already rejects this
+                # combination before params reach here, but `_amessages` can be called directly
+                # (e.g. in tests), so guard here too rather than silently dropping `stream`.
+                msg = "stream is not supported for output_format"
+                raise ValueError(msg)
+            native_kwargs = params.model_dump(
+                exclude_none=True, exclude={"output_format", "stream", "betas", "context_management"}
```
Comment (HareeshBahuleyan) [inline src/any_llm/providers/meta/meta.py:143]: ### [P1] `prompt_cache_key` breaks the Messages path

The provider advertises `PROMPT_CACHE_KEY_SUPPORT = "supported"`, but `_amessages()` forwards `prompt_cache_key` into the Anthropic SDK. Anthropic SDK 0.83 does not accept that argument:

```text
TypeError: AsyncMessages.create() got an unexpected keyword argument 'prompt_cache_key'
```

Meta documents `prompt_cache_key` for Chat Completions and Responses, but not for Messages. The Messages implementation should exclude it and raise a clear endpoint-specific unsupported-parameter error. It should remain supported for the OpenAI-compatible paths.

File: src/any_llm/providers/meta/meta.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            )
+            native_kwargs.update(kwargs)
+            if is_structured_output_type(params.output_format):
+                return await self._anthropic_client.messages.parse(output_format=params.output_format, **native_kwargs)
+            message = await self._anthropic_client.messages.create(
+                output_config=cast("Any", params.output_format), **native_kwargs
+            )
+            return self._convert_native_message_to_response(message)
+
+        api_kwargs = params.model_dump(exclude_none=True, exclude={"betas", "context_management"})
```
Comment (HareeshBahuleyan) [inline src/any_llm/providers/meta/meta.py:153]: ### [P2] Unsupported Messages parameters are passed through

The native pass-through forwards `stop_sequences`, `top_k`, and top-level `cache_control`.

Meta's Messages documentation explicitly says:

- `stop_sequences` and `top_k` return HTTP 400.
- Unknown top-level fields return HTTP 400.
- Top-level `cache_control` is not documented.

Therefore, the claim that `cache_control` “survives the round trip” is misleading. These parameters should be explicitly rejected, or the documentation should clearly describe the limited Messages compatibility.

**Comment by HareeshBahuleyan** [comment]
Maybe we could simplify this provider by keeping the inherited Chat Completions and Responses support, while removing the native Messages implementation?

Meta documents Responses as the full-featured interface for agent workflows, including reasoning replay, tool loops, search grounding, files, and server-managed history. Chat Completions and Responses use the same `AsyncOpenAI` client and require little provider-specific code.

Most of the maintenance complexity comes from the second `AsyncAnthropic` client, base URL derivation, unsupported-parameter filtering, streaming event conversion, and additional tests. Basic `.messages()` usage could still rely on any-llm’s existing Messages-to-Completions bridge.

The tradeoff is losing native Anthropic fidelity and direct Claude Code compatibility. Unless Claude Code or native Anthropic blocks are explicit requirements, keeping only the OpenAI-compatible paths would retain advanced agent support while substantially reducing maintenance.

File: src/any_llm/providers/meta/meta.py (hunk 6 above)
Comment (tbille) [reply src/any_llm/providers/meta/meta.py:143]: Fixed in fb3e44c: `_amessages` now rejects `prompt_cache_key` client-side with `UnsupportedParameterError` before it ever reaches the Anthropic SDK call, instead of letting `messages.create()` raise a raw `TypeError`. It remains supported (unchanged) on the OpenAI-compatible Chat Completions/Responses paths, since `PROMPT_CACHE_KEY_SUPPORT = "supported"" only gates the generic base-layer check, not this provider's Messages-specific handling. Added parametrized test coverage.

File: src/any_llm/providers/meta/meta.py (hunk 7 above)
Comment (tbille) [reply src/any_llm/providers/meta/meta.py:153]: Fixed in fb3e44c: `stop_sequences` and `top_k` are now rejected the same way as `prompt_cache_key` above, with `UnsupportedParameterError`, matching the existing `context_management`/`betas` guard.

On `cache_control`: you're right that the docstring overclaimed it. Meta's docs don't explicitly confirm or reject it for Messages (unlike `stop_sequences`/`top_k`, which are called out by name), so I reworded the docstring to stop asserting it survives, rather than adding a rejection I can't back with a docs citation either way. Given it's a real top-level kwarg on the Anthropic SDK's own `messages.create()` (used for prompt caching, not something Meta specifically excluded in its documented field list), blocking it outright felt like the wrong kind of guess in the other direction. Whoever runs the live-verification pass can confirm either way and I'll follow up.

**Review by HareeshBahuleyan — APPROVED** [review]
